shared/controlflow/codeql/controlflow/BasicBlock.qll

```diff
@@ -0,0 +1,258 @@
+     */
+    predicate postDominates(BasicBlock bb) {
+      this.strictlyPostDominates(bb) or
+      this = bb
+    }
+
+    /** Holds if this basic block is in a loop in the control flow graph. */
+    predicate inLoop() { this.getASuccessor+() = this }
+
+    /** Gets a textual representation of this basic block. */
+    string toString() { result = this.getFirstNode().toString() }
+  }
+
+  cached
+  private module Cached {
+    /**
+     * Internal representation of basic blocks. A basic block is represented
+     * by its first CFG node.
+     */
+    cached
+    newtype TBasicBlock = TBasicBlockStart(Node cfn) { startsBB(cfn) }
+
+    /**
+     * Holds if the first node of basic block `succ` is a control flow
+     * successor of the last node of basic block `pred`.
+     */
+    private predicate succBB(BasicBlock pred, BasicBlock succ) { pred.getASuccessor(_) = succ }
+
+    /** Holds if `dom` is an immediate dominator of `bb`. */
+    cached
+    predicate bbIDominates(BasicBlock dom, BasicBlock bb) =
+      idominance(entryBB/1, succBB/2)(_, dom, bb)
+
+    /** Holds if `pred` is a basic block predecessor of `succ`. */
+    private predicate predBB(BasicBlock succ, BasicBlock pred) { succBB(pred, succ) }
+
+    /** Holds if `bb` is an exit basic block that represents normal exit. */
+    private predicate exitBB(BasicBlock bb) { nodeIsPostDominanceExit(bb.getANode()) }
+
+    /** Holds if `dom` is an immediate post-dominator of `bb`. */
+    cached
+    predicate bbIPostDominates(BasicBlock dom, BasicBlock bb) =
+      idominance(exitBB/1, predBB/2)(_, dom, bb)
+  }
+
+  private import Cached
+
+  /** Holds if `cfn` starts a new basic block. */
+  private predicate startsBB(Node cfn) {
+    not exists(nodeGetAPredecessor(cfn, _)) and exists(nodeGetASuccessor(cfn, _))
+    or
+    nodeIsJoin(cfn)
+    or
+    nodeIsBranch(nodeGetAPredecessor(cfn, _))
+    or
+    // In cases such as
+    //
+    // ```rb
+    // if x or y
+    //     foo
+    // else
+    //     bar
+    // ```
+    //
+    // we have a CFG that looks like
+    //
+    // x --false--> [false] x or y --false--> bar
+    // \                    |
+    //  --true--> y --false--
+    //            \
+    //             --true--> [true] x or y --true--> foo
+    //
+    // and we want to ensure that both `foo` and `bar` start a new basic block.
+    exists(nodeGetAPredecessor(cfn, any(SuccessorType s | successorTypeIsCondition(s))))
+  }
+
+  /**
+   * Holds if `succ` is a control flow successor of `pred` within
+   * the same basic block.
+   */
+  predicate intraBBSucc(Node pred, Node succ) {
+    succ = nodeGetASuccessor(pred, _) and
+    not startsBB(succ)
+  }
+
+  /**
+   * Holds if `bbStart` is the first node in a basic block and `cfn` is the
+   * `i`th node in the same basic block.
+   */
+  private predicate bbIndex(Node bbStart, Node cfn, int i) =
+    shortestDistances(startsBB/1, intraBBSucc/2)(bbStart, cfn, i)
```

Comment:
As I mentioned the other day, I don't think `shortestDistances` is optimal. IIRC it can be beat with a QL recursion, but it's something that'll need fresh measurements on some large examples. I expect the C++ implementation to be in the lead in terms of performance on this particular predicate, and C++ uses a straightforward QL recursion.

Reply:
Since that requires some benchmarking I'd prefer to leave that for a future PR.